crates/biome_js_analyze/src/analyzers/nursery/no_restricted_imports.rs

```diff
@@ -0,0 +1,74 @@
+use biome_analyze::context::RuleContext;
+use biome_analyze::{declare_rule, Ast, Rule, RuleDiagnostic, RuleSource};
+use biome_console::markup;
+use biome_deserialize_macros::Deserializable;
+use biome_js_syntax::inner_string_text;
+use biome_rowan::TextRange;
+use rustc_hash::FxHashMap;
+use serde::{Deserialize, Serialize};
+
+use super::use_nodejs_import_protocol::AnyJsImportLike;
+
+declare_rule! {
+    /// Disallow specified modules when loaded by import or require.
+    ///
+    /// https://eslint.org/docs/latest/rules/no-restricted-imports
+    ///
+    /// ## Options
+    ///
+    /// ```json
+    /// {
+    ///     "noRestrictedImports": {
+    ///         "options": {
+    ///             "paths": {
+    ///                 "lodash": "Using lodash is not encouraged",
+    ///                 "underscore": "Using underscore is not encouraged"
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    pub(crate) NoRestrictedImports {
+        version: "next",
+        name: "noRestrictedImports",
+        source: RuleSource::Eslint("no-restricted-imports"),
+        recommended: false,
+    }
+}
+
+/// Options for the rule `noRestrictedImports`.
+#[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct RestrictedImportsOptions {
+    /// A list of names that should trigger the rule
+    #[serde(skip_serializing_if = "FxHashMap::is_empty")]
+    paths: FxHashMap<String, String>,
+}
+
+impl Rule for NoRestrictedImports {
+    type Query = Ast<AnyJsImportLike>;
+    type State = (TextRange, String);
+    type Signals = Option<Self::State>;
+    type Options = RestrictedImportsOptions;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let module_name = ctx.query().module_name_token()?;
+        let inner_text = inner_string_text(&module_name);
+
+        ctx.options()
+            .paths
+            .get(inner_text.text())
+            .map(|message| (module_name.text_trimmed_range(), message.to_string()))
```

Comment:
The options are also available inside the `diagnostic` function, which means that inside `run`, we can return a `TextRange` and the index of where we found the banned module, without allocating a second string via `message.to_string`. 

Plus, inside `diagnostic` we only need reference to the string, when we do the printing.

Reply:
I didn't understand this. The index of a hash map has a type of string, and we don't want to return a string?